presto-main/src/main/java/com/facebook/presto/util/MergeSortedPages.java

```diff
@@ -100,6 +101,7 @@ private MergeSortedPages() {}
 
             boolean finished = !pageWithPositionOptional.isPresent();
             if (finished && pageBuilder.isEmpty()) {
+                memoryContext.close();
```

Comment:
Is it guaranteed this will be called, like `finally`? If the resulting `WorkProcessor` is fed to some other work processor that happens to be implementing LIMIT, it will stop reading from the returned WP and local memory context won't be closed.

Since `WorkProcessor` embeds a closeable resource, maybe it should implement `AutoCloseable` itself. API-wise, this could look like:
```
return pageWithPositions.transform(pageWithPositionOptional -> {
    ....
    })
   .onClose(memoryContext::close);  // or .closing(memoryContext)
```

Reply:
> Since WorkProcessor embeds a closeable resource, maybe it should implement AutoCloseable itself. API-wise, this could look like:

Let's not introduce this complexity when it's not needed yet. This should be addressed in the next step when operators are combined together. The idea is that there will be some:
```
CloseableWorkProcessor<Page> Operator/OperatorFactory#transform(WorkProcessor<Page> source)
```
method. Then `close` will be called like it's currently called for operators when upstream operator (like `LIMIT`) finishes early. Such close should dereference objects and close memory contexts. After `close` any calls to work processor is no longer valid.